### PR TITLE
WIP tests(upgrade): test upgrade path from Elgg 2.3 to current pull request

### PR DESCRIPTION
**Pull request by hypeJunction**

### CODE REVIEW

**Comment by hypeJunction** [comment]
This doesn't really work, but you get the vibe.

**Comment by mrclay** [comment]
Yeah, this would be a great help even if it doesn't run the async migrations.

File: .travis.yml [hunk 1]
```diff
@@ -112,6 +112,25 @@ matrix:
...
+      env: JOB_NAME=upgrade_from_Elgg2 VARIA=true
+      before_install:
+        - phpenv config-rm xdebug.ini
+        - composer config -g github-oauth.github.com ${GITHUB_TOKEN}
+      install:
+        - git clone https://github.com/$TRAVIS_REPO_SLUG.git $TRAVIS_REPO_SLUG
+        - cd $TRAVIS_REPO_SLUG
+        - git checkout 2.3
+        - composer travis:install-with-mysql
+        - git checkout -qf $TRAVIS_COMMIT
```
Comment (mrclay) [inline .travis.yml]: What does this line do?

File: .travis.yml [hunk 2]
```diff
@@ -112,6 +112,26 @@ matrix:
...
+        - phpenv config-rm xdebug.ini
+        - composer config -g github-oauth.github.com ${GITHUB_TOKEN}
+      install:
+        - git clone https://github.com/$TRAVIS_REPO_SLUG.git $TRAVIS_REPO_SLUG
+        - cd $TRAVIS_REPO_SLUG
+        - git checkout 2.3
+        - composer travis:install-with-mysql
+        - git fetch origin
+        - git checkout -qf $TRAVIS_COMMIT
+        - php -f ./upgrade.php
```
Comment (mrclay) [inline .travis.yml]: Application::upgrade needs to check for CLI and not require the GET input.
forward() should send an instruction to stdout I guess?

File: .travis.yml [hunk 3]
```diff
@@ -112,6 +112,26 @@ matrix:
...
+      script:
+        - ./vendor/bin/phpunit
+        - php -f ./engine/tests/suite.php
```
Comment (mrclay) [inline .travis.yml]: We should boot up the PHP server, too.

File: .travis.yml (hunk 2 above)
Comment (hypeJunction) [inline .travis.yml]: `elgg-cli` would help here
Elgg should really stop being an HTTP diva and think about other kernels

File: .travis.yml (hunk 3 above)
Comment (hypeJunction) [inline .travis.yml]: I am trying to get the correct head checked out first before anything runs

**Comment by mrclay** [comment]
This feels close. I guess now we have to decide how much dummy content to inject so that that upgrader doesn't just sail past problem areas. BUT we can now add that with PRs that have upgrade libs.

**Comment by mrclay** [comment]
I suppose those *could* be legit bugs in our simpletests

**Comment by hypeJunction** [comment]
No, we need to also migrate the site url

**Comment by mrclay** [comment]
:clinking_glasses:

**Comment by hypeJunction** [comment]
🍾

**Comment by hypeJunction** [comment]
Want to polish it up? It's 11pm and I should be out drinking.

**Comment by mrclay** [comment]
Yep I'll do that.

**Comment by hypeJunction** [comment]
Well, this was fun 🥇

**Comment by hypeJunction** [comment]
> Yeah, this would be a great help even if it doesn't run the async migrations.

One of these days, I will put together some seeders, so we can start testing async migrations as well.

File: engine/classes/Elgg/Config.php [hunk 4]
```diff
@@ -289,6 +292,77 @@ public function loadSettingsFile() {
 	}
 
 	/**
+	 * Pre 3.0 certain config values were store in the database
+	 * To allow for a smooth upgrade, we will attempt to copy these settings from the database
+	 * to settngs.php, where these values will be read from by 3.0 codebase
+	 *
+	 * @return void
+	 */
+	public function migrateDbSettings($path) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Config.php]: Ideally, we should make these easy to implement. Perhaps some directly in `/install` where we execute them all.

Comment (hypeJunction) [inline engine/classes/Elgg/Config.php]: Perhaps name them after config values:

```php
if (!$CONFIG->dataroot) {
   require_once $path . '/install/config/dataroot.php';
}
```

**Comment by mrclay** [comment]
Replaced by #11113

**Pull request closed without merging**